src/Chainweb/Pact/Types.hs

```diff
@@ -105,4 +105,16 @@ newtype GasId = GasId PactId deriving (Eq, Show)
 
 type TransactionM p a = ReaderT (CommandEnv p) IO a
 
+data Rewind a
+    = DoRewind !a
+    | NoRewind !a
+    | Genesis {-# UNPACK #-} !cid
+    deriving (Functor, Applicative, Foldable, Traversable, Monad)
+
+instance HasChainId a => HasChainId (Rewind a) where
+    _chainId = \case
+      DoRewind !a -> _chainId a
+      NoRewind !a -> _chainId a
+      Genesis !cid -> cid
+
```

Comment:
Why do you need Functor, Applicative, Foldable, Traversable and Monad instances?

Reply:
Cautionary - i don't know where I'll be going so I want to know i have everything I need and I'll pare it down as I work the PR. It's a WIP draft!